sourceaddrs/source_registry_final.go

```diff
@@ -62,3 +95,8 @@ func (s RegistrySourceFinal) FinalSourceAddr(realSource RemoteSource) RemoteSour
 	// paths together, so we can just delegate to our unversioned equivalent.
 	return s.Unversioned().FinalSourceAddr(realSource)
 }
+
+// finalRegistrySourcePattern is a non-exhaustive regexp which looks only for
+// the expected three components of a RegistrySourceFinal string encoding: the
+// package address, version, and subpath.
+var finalRegistrySourcePattern = regexp.MustCompile(`^(.+)@(.+)//(.+)$`)
```

Comment:
Although not true for all of the single-file addresses you've been dealing with in your current work, it's possible to have a "final address" without a sub-path portion if it refers to the root directory of the registry module package.

Therefore I think the part of this pattern starting with `//` ought to be optional, although I suspect doing that will make the pattern highly ambiguous (it will essentially just be "is there at least one at sign?") so we might also need to extend the "looks like final registry source" function to actually try parsing the version part as a version number, before concluding that it's safe to discard it and see if what's left is a valid non-final registry source. 🤔

Reply:
Ah, thanks for spotting this oversight. I've updated the pattern and the code using it, with one caveat:

> we might also need to extend the "looks like final registry source" function to actually try parsing the version part as a version number

I'm not sure about this part. Is this from a "what if?" cautious stance, or are there ambiguities in the final source address format I'm not seeing that make it necessary?

I intend the regexp only to help split the three (or two!) components of the final registry source, and lean on the registry address parser (ignoring the version) to check if the string looks like a final registry source. I don't think it's possible for any other source address type to match the regexp, and then have its package and subpath parse as a registry source address.

The reason I'm asking is because making an address like `example.com/foo/bar/baz@1.0.x` fail to match `looksLikeFinalRegistrySource` means that it will instead be interpreted as a remote source. As a result we won't see a useful error explaining that the version is invalid.

Reply:
I must confess that I am coming at this from a cautious standpoint because in the similar code over in Terraform Core we initially had a not-quite-tight-enough definition of "is a registry source address" which caused Terraform to misinterpret some addresses that were intended to be remote source addresses, causing confusing error messages. That was hard to resolve retroactively because we needed to avoid breaking the handling of any addresses which the author _did_ intend to be a registry source address. :grimacing:  (There's some commentary resulting from that earlier mess [in the Terraform Core `addrs` package](https://github.com/hashicorp/terraform/blob/4b1c3457a4090235645b699ec360ff9a7a2f4838/internal/addrs/module_source.go#L83-L96).

With that said, I would be happy if we can convince ourselves that the concern doesn't apply here by adding additional test cases, rather than adding additional code.

---

You have pointed out something in your response which I think is the same potential problem viewed from the opposite angle: the goal of this function is to try to differentiate between user intent to write a registry source and user intent to write a remote source. If we misinterpret the user's intent in either direction then the result will be a confusing error message. I'm worrying about misclassifying a remote source as a registry source, and you're worrying about the converse, but a misclassification in either direction is equally confusing.

On further reflection, I think we might be saved here by this implementation being only a subset of what `go-getter` would accept, meaning that we don't have quite as many highly-ambiguous forms as Terraform itself currently needs to deal with. I used the tests for `ParseSource` to remind myself what's included in the subset that `go-slug` implements and I think the only ones that seem worthy of some careful scrutiny are these shorthand `github.com` and `gitlab.com` variants, which are syntactically close to the registry address syntax:

https://github.com/hashicorp/go-slug/blob/aeb5117bdbb942fec6a3308de57ecf96191316c9/sourceaddrs/source_test.go#L213-L273

So some examples of what might be misclassified by an unsufficiently-specific rule could include:

* `github.com/foo/bar/baz@1.0.0` (a directory called `baz@1.0.0` inside a GitHub repository)
* `github.com/foo/bar//baz@1.0.0` (the previous but written using the explicit sub-path syntax)

The last case in the range I linked above is also pretty :grimacing: -- `gitlab.com` is _both_ a Git repository host and a Terraform module registry, and so the handling of that is _unavoidably_ ambiguous. If we are parsing something that could be either then it seems we prefer the module registry interpretation and require `git::` prefix to access the other interpretation. So I suppose we should make sure we still resolve that ambiguity in the same way after the registry interpretation has becomes a "final" registry address.

Reply:
I think our disconnect here stems from this:

> the goal of this function is to try to differentiate between user intent to write a registry source and user intent to write a remote source

My intent with this function wasn't intended to handle user input. It's intended to parse the encoded output of previously-parsed user input. But since there's nothing _stopping_ a user of this library from running it on user input, I can see the concern now.

As of my last commit, the first example above has a direct equivalent in the tests, which I think works as you'd expect: https://github.com/hashicorp/go-slug/blob/7ff3155b01cf46c00449e272b8dc90b1e80c8c06/sourceaddrs/source_final_test.go#L134-L137

The second example above currently doesn't parse at all, either with `ParseSource` or `ParseFinalSource`:

> `source_test.go:428: unexpected error: invalid remote source address "gitlab.com/hashicorp/go-slug//bleep/bloop": invalid sub-path: must be slash-separated relative path without any .. or . segments`

I confess I'm rather turned around at this point, and I'm not sure what changes you're recommending. If you could let me know what your current thinking is I'll be happy to implement whatever you suggest.

Reply:
Those examples failing to parse at all is also a fine outcome. I was concerned only that they might be misunderstood as being of the wrong type.

I'm going to now re-read the tests as you currently have them with fresh eyes and brain.

Reply:
I think the current suite of tests you have is sufficient to show that any ambiguities in the regex are only hypothetical problems, and not actual problems. :+1: